File: internal/library/anime/family_tree.go
package anime

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
)

type OfflineDBEntry struct {
	Sources      []string            `json:"sources"`
	Title        string              `json:"title"`
	Type         string              `json:"type"`
	Relations    []OfflineDBRelation `json:"relations"`
	AnilistID    *int                `json:"anilist_id,omitempty"`
	MALID        *int                `json:"mal_id,omitempty"`
}

type OfflineDBRelation struct {
	RelationType string `json:"relationType"`
	ID           int    `json:"id"`
}

type AnimeOfflineDatabase struct {
	Data struct {
		Anime []OfflineDBEntry `json:"anime"`
	} `json:"data"`
}

var offlineDB *AnimeOfflineDatabase

func LoadAnimeOfflineDatabase(jsonPath string) error {
	f, err := os.Open(filepath.Clean(jsonPath))
	if err != nil {
		return fmt.Errorf("failed to open anime-offline-database.json: %w", err)
	}
	defer f.Close()
	dec := json.NewDecoder(f)
	var db AnimeOfflineDatabase
	if err := dec.Decode(&db); err != nil {
		return fmt.Errorf("failed to decode anime-offline-database.json: %w", err)
	}
	offlineDB = &db
	return nil
}

// FindFamilyTree returns the canonical, chronological, alternative, and character-used-from relations for the given AniList ID
func FindFamilyTree(anilistID int) (canonical []OfflineDBEntry, chronological []OfflineDBEntry, alternatives []OfflineDBEntry, charactersFrom []OfflineDBEntry, err error) {
	if offlineDB == nil {
		err = fmt.Errorf("offline database not loaded")
		return
	}
	var thisEntry *OfflineDBEntry
	for i, entry := range offlineDB.Data.Anime {
		if entry.AnilistID != nil && *entry.AnilistID == anilistID {
			thisEntry = &offlineDB.Data.Anime[i]
			break
		}
	}
	if thisEntry == nil {
		err = fmt.Errorf("anime not found in offline database: %d", anilistID)
		return
	}
	for _, rel := range thisEntry.Relations {
		for i, e := range offlineDB.Data.Anime {
			if e.AnilistID != nil && *e.AnilistID == rel.ID {
				switch rel.RelationType {
				case "Prequel", "Sequel":
					canonical = append(canonical, offlineDB.Data.Anime[i])
				case "Alternative version", "Alternative setting":
					alternatives = append(alternatives, offlineDB.Data.Anime[i])
				case "Character":
					charactersFrom = append(charactersFrom, offlineDB.Data.Anime[i])
				case "Side story", "Summary", "Other":
					chronological = append(chronological, offlineDB.Data.Anime[i])
				}
			}
		}
	}
	return
}

File: seanime-web/src/lib/family-tree.ts
import path from "path";
import { promises as fs } from "fs";

const DB_PATH = path.resolve(process.cwd(), "external/anime-offline-database.json");

export interface FamilyTreeEntry {
  id: number;
  title: string;
  type?: string;
  relationType?: string;
}

export interface FamilyTreeData {
  canonical: FamilyTreeEntry[];
  chronological: FamilyTreeEntry[];
  alternatives: FamilyTreeEntry[];
  charactersFrom: FamilyTreeEntry[];
}

/**
 * Looks up a media entry by local id, extracts AniList ID, and fetches family tree data from AniList.
 * @param mediaId Local media id (string or number)
 * @returns FamilyTreeData or throws error
 */
export async function getFamilyTreeForMediaId(mediaId: string | number): Promise<FamilyTreeData> {
  let db: any;
  try {
    const data = await fs.readFile(DB_PATH, "utf-8");
    db = JSON.parse(data);
  } catch (e) {
    throw new Error("Could not load offline database");
  }

  // Find the anime entry in offline DB by mediaId
  const entry = (db.data || db.anime)?.find((a: any) => String(a.id) === String(mediaId));
  if (!entry) throw new Error("Not found");

  // Extract AniList ID from sources
  let aniListId: string | undefined = undefined;
  if (Array.isArray(entry.sources)) {
    for (const src of entry.sources) {
      const match = src.match(/anilist.co\/anime\/(\d+)/);
      if (match) {
        aniListId = match[1];
        break;
      }
    }
  }
  if (!aniListId) {
    throw new Error("AniList ID not found in sources");
  }

  // Query AniList GraphQL API for relations
  const query = `
    query ($id: Int!) {
      Media (id: $id) {
        id
        title { romaji }
        relations {
          edges {
            relationType
            node {
              id
              title { romaji }
              type
            }
          }
        }
      }
    }
  `;
  let relationsData: any[] = [];
  try {
    const resp = await fetch("https://graphql.anilist.co", {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ query, variables: { id: Number(aniListId) } })
    });
    const json = await resp.json() as {
      data?: {
        Media?: {
          relations?: {
            edges?: any[]
          }
        }
      }
    };
    relationsData = json.data?.Media?.relations?.edges || [];

  } catch (e) {
    throw new Error("Failed to fetch from AniList");
  }

  // Map AniList relation types to FamilyTreeData
  function filterByType(typeArr: string[]) {
    return relationsData.filter((edge: any) => typeArr.includes(edge.relationType)).map((edge: any) => ({
      id: edge.node.id,
      title: edge.node.title.romaji,
      type: edge.node.type,
      relationType: edge.relationType
    }));
  }
  const familyTreeData: FamilyTreeData = {
    canonical: filterByType(["PREQUEL", "SEQUEL", "PARENT", "CHILD"]),
    chronological: filterByType(["ADAPTATION", "SOURCE", "SUMMARY", "SPIN_OFF"]),
    alternatives: filterByType(["ALTERNATIVE", "SIDE_STORY", "CHARACTER"]),
    charactersFrom: filterByType(["CHARACTER"]),
  };

  return familyTreeData;
}
